java-checks/src/main/java/org/sonar/java/checks/TooManyParametersCheck.java

```diff
@@ -66,33 +63,60 @@ public class TooManyParametersCheck extends BaseTreeVisitor implements JavaFileS
     "javax.ws.rs.PUT",
     "javax.ws.rs.PATCH",
     "org.springframework.beans.factory.annotation.Autowired",
-    "javax.inject.Inject"
-  );
+    "javax.inject.Inject",
+    "org.springframework.context.annotation.Bean",
+    "io.micronaut.http.annotation.Get",
+    "io.micronaut.http.annotation.Post",
+    "io.micronaut.http.annotation.Put",
+    "io.micronaut.http.annotation.Delete",
+    "io.micronaut.http.annotation.Options",
+    "io.micronaut.http.annotation.Patch",
+    "io.micronaut.http.annotation.Head",
+    "io.micronaut.http.annotation.Trace");
+
+  // if a class is annotated as one of these types, its constructor should be ignored if it's the only constructor
+  private static final List<String> CLASS_ANNOTATION_CONSTRUCTOR_EXCEPTIONS = Arrays.asList(
+    "org.springframework.stereotype.Component",
+    "org.springframework.context.annotation.Configuration",
+    "org.springframework.stereotype.Service",
+    "org.springframework.stereotype.Repository");
 
   @Override
-  public void scanFile(JavaFileScannerContext context) {
-    this.context = context;
-    scan(context.getTree());
+  public List<Tree.Kind> nodesToVisit() {
+    // This rule has the following exceptions: RECORD, ANNOTATION_TYPE
+    return Arrays.asList(Tree.Kind.CLASS, Tree.Kind.INTERFACE, Tree.Kind.ENUM);
   }
 
   @Override
-  public void visitMethod(MethodTree tree) {
-    super.visitMethod(tree);
-    if (isOverriding(tree) || usesAuthorizedAnnotation(tree)) {
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+
+    // this is true only if the class has unknown annotations, or if it has a single constructor and
+    // it is annotated with one of CLASS_CONSTRUCTOR_WHITELIST
```

Comment:
`CLASS_ANNOTATION_CONSTRUCTOR_EXCEPTIONS` instead of `CLASS_CONSTRUCTOR_WHITELIST`